### ISSUE TITLE
Unable to press NVDA key when running in a VM on Mac host

### ISSUE THREAD

**Issue by nvaccessAuto** [nvaccess/nvda#225]
**Reported by vtsaran on 2008-11-10 23:23**
When running NVDA inside virtual machines (particularly on the Mac) it is practically impossible to use INSERT, CAPSLOCK and CONTEXTMENU keys, present on most Windows keyboards. This makes it impossible to modify most of NVDA settings or access NVDA menus.
It would be desirable, therefore, to have a way to assign different key combination for NVDA key.
**Blocking #347**

**Comment by feerrenrut** [comment]
For anyone coming across this issue, some guidance on how to setup your NVDA key on a Mac while using a VM can be found in the following article: https://www.marcozehe.de/2015/06/07/how-to-map-your-macs-capslock-key-to-a-nvda-or-jaws-key-in-a-windows-virtual-machine/

**Comment by bhavyashah** [comment]
@feerrenrut The original use case that vtsaran provided seems to be partially or fully resolved now. Furthermore, since a hidden option for such configurability was the agreed approach, an average NVDA user might hardly benefit from such a feature. Thus, should we accept this feature request, assign it a P4, or close it as a wontfix?

**Comment by feerrenrut** [comment]
I'm not aware that this is resolved, I pointed to a possible workaround but we have no confirmation that this works. I think its a good idea to try support web developers who wish to test with NVDA and happen to be running a mac, ultimately this will benefit NVDA users. As such we would be willing to discuss potential changes proposed by external developers in order to cater for this case.

Given the user-story "As a web-developer testing for A11Y using NVDA in a VM on my Mac, I want to be able to reassign the NVDA key because my keyboard does not have the required keys and the VM I use does not allow me to remap keys". I think it is quite unlikely to be prioritised as a feature to include in  a project. This might be different if we knew that there were a lot of people testing for A11Y this way / or knowing if the workaround did not work.

I'll label this as a feature request. Happy to discuss this with anyone who wishes to implement it.

**Comment by a-singer** [comment]
This is a copy of my use case from a closed issue listed above. I add it to the Mac and accessibility testing use cases above.

I am using NVDA from iOS through a VNC application. I find that this works fairly well except that the application will not properly pass through the capslock or insert key as a modifier. I am, therefore, unable to use the NVDA commands without another program being involved to modify the keyboard. I have used autohotkey to map the right alt key to the insert key. Sharpkeys and other registry-based keyboard mapping applications don’t work. I would like the ability to have the NVDA key set to any key on the keyboard that Windows can see. I have seen keyboards with, for example, the ` key next to the space and there is always, of course, one of the alt keys. Given that NVDA, unlike Jaws, will happily work with VNC input, this may be worth doing as I have yet to see a VNC application which sends the capslock key as a modifier properly.

**Comment by DrSooom** [comment]
### Actual NVDA modifier keys in NVDA 2018.3.2:
+ Insert (E0_52)
+ NumInsert/Num0 (00_52)
+ CapsLock (00_3A)
### Expected NVDA modifier keys:
+ Insert (E0_52)
+ NumInsert/Num0 (00_52)
+ CapsLock (00_3A)
+ NumLock (00_45)
+ ScrollLock (00_46)
+ Application (E0_5D)
+ PrtSc (E0_37)
#### Reasons:
+ In my opinion adding those four new keys should be quite easy to implement, as three of them – as far as I know – aren't used for hotkeys.
+ But according to the PrtSc key the chapter "4.2.1. The NVDA Modifier Key" in the NVDA User Guide has to contain also a little note, that if this key is used as a NVDA modifier key, ALT+ and Win+ only works correctly when pressing NVDA+F2 (Pass next key through) first.
+ The DEL key is too risky and the ESC key is already used for Windows-based hotkeys like ALT+ESC, CTRL+ESC and CTRL+Shift+ESC.
+ Well, and F1 and F12 might also be used for hotkeys, so I don't think that it make sense to use them as a NVDA modifier key as well.
+ And note that there are hundreds of different keyboard layouts out there. The German one for example has no right ALT key – there is a key called AltGr (combination of ALT+CTRL). And some keyboards no longer have two WIN keys.
+ Using both Shift keys, both ALT key, both WIN keys and the left CTRL key as a NVDA modifier key isn't useful at all, because there are too much hotkeys which need these keys.
+ At least we only could add the right CTRL key (E0_1D) as a NVDA modifier key, as VirtualBox uses this key for toggling the keyboard command back ´to the host.

**Comment by a-singer** [comment]
> #### Reasons:
> * In my opinion adding those four new keys should be quite easy to implement, as three of them – as far as I know – aren't used for hotkeys.
> * And note that there are hundreds of different keyboard layouts out there. The German one for example has no right ALT key – there is a key called AltGr (combination of ALT+CTRL). And some keyboards no longer have two WIN keys.


I agree with what you say above, and this is just why I wish to have the key configurable by the user, rather than even an expanded list. Obviously, I wouldn't mind a hidden option either with or without an expanded list. My difficulty is just the very large number of keyboard layouts and use cases out there. None of the keys you suggest above would work for my use case, since they wouldn't be sent as modifiers on VNC. Similarly, the key that I'm using, the right alt, wouldn't work on the German layout, for the reason you mention. With a configurable key, the user can decide what to do depending on the unique circumstances.
Thanks,

**Comment by DrSooom** [comment]
To prevent user from misconfiguration a list would be the easiest solution via the GUI.

But I also can understand, that for some users ESC (00_01) or F1 (00_3B) to F12 (00_58) would be a better NVDA modifier key. But if we implement F1 to F12 as an available option, we need to implement routines to automatically changes some input gestures. And that makes this so complicated.

Furthermore, on the German keyboard layout there are only three additional keys which can be used as a NVDA modifier key: + (00_1B), # (00_2B) and < (00_56. But I guess that there would be more available keys on an EN-US layout keyboard, which also can be used as a NVDA modifier key. So in the end we need a keyboard layout depending list of available keys here.

CC: @feerrenrut and @ehollig: Your thoughts about this?

**Comment by a-singer** [comment]
> To prevent user from misconfiguration a list would be the easiest solution via the GUI.
 

That is certainly true, but a documented hidden option, as suggested by @jteh would have the best of both worlds. I respectfully agree that it should not be possible, through user error, to end up with no NVDA key or a mess which causes commands to be impossible to give to the PC through any keyboard.

 
> So in the end we need a keyboard layout depending list of available keys here.


Wouldn't this be very complicated, if I may ask? It would have to be checked, if new or custom layouts were used issues would come up... It just seems to me more difficult than a general list and a way in which those unusual users who need an extra key can add one.

**Comment by DrSooom** [comment]
The list of available keys has two parts: a global firm one and a localized one.
+ Firm keys: Insert, NumInsert, CapsLock, NumLock, ScrollLock, Application, PrtSc
+ If desired those additional firm keys: ESC, RightSCTRL, RightWIN and F1 to F12
Note that PrtSc, ESC, RightCTRL, RighWIN and F1 to F12 are already used for hotkeys. This must be mentioned in the NVDA User Guide.
+ Localized German (DE-DE and DE-AT) keys: <, +, # and - (I wouldn't prefer the last one.)

German (DE-DE and DE-AT) keyboard layout in one single row:
> ^1234567890ß´ qwertzuiopü+ asdfghjklöä# <yxcvbnm,.-

In total 48 characters/keys, but A to Z and 1 to 0 cannot be used as a NVDA modifier key. And it's the same with ß, ä, ö, ü, , (comma), . (dot) and – in my opinion – - (dash). ^ and ´ have a special behaviour on the German keyboard, so they can't be used either. In other words: only 3 (or 4) of the 48 keys can be used for the German localized list of available keys.

Furthermore I'm asking me if NVDA is able to read the current used keyboard layout from the OS settings, like with the language. This could help here a little bit, otherwise we would need an additional dropdown list for the keyboard layouts as well.

**Comment by Adriani90** [comment]
@feerrenrut now that emulated system keys can be assigned in the input gesture dialog, I think it is worthy to consider adding this feature to NVDA. What do you think?

**Comment by Qchristensen** [comment]
I had a request from a user to be able to assign ScrollLock as the NVDA key.  I'm not sure how many keyboards have an easier to access scroll lock key as opposed to caps lock or insert, but the user made the assertion that scroll lock is not widely used, and that part I can't argue with.

**Comment by NireBryce** [comment]
Prefacing this with that I'm not visually impaired but I have pretty strong dyslexia.

I can't contribute to this project at the moment, but this is sort of a glaring issue where every few years I try to hop into screen readers (a boon! if I could control them enough to make them work for me), discover this has not been fixed *for the ten years I have been trying*, give up and leave.  How many others are like this who aren't forced to deal with this decision out of necessity or habit?

There are no windows screen readers that let me bind their interact key to something that's not capslock or insert.  It's user hostile to not have this something that can at least be reconfigured.  

application, right ctl, right win, would be better options if we're restricted to the keys posted a few posts up.

This has been an issue since 2008.  this is *accessibility software*, and yet the answer is 'go through a lot of poweruser steps to rebind another key to capslock/insert'.  That is, IMO, unacceptable from an a11y standpoint.

How are we still here? It seems like an oversight to expect every user, sighted or not, to settle for hardcoded controls, especially hardcoded controls many of us try to avoid in general due to their togglable nature.

This is frustrating, but at least I don't *need* screen readers while being this frustrated.

**Comment by feerrenrut** [comment]
The initial here describes a use-case of using NVDA via a virtual machine while on Apple hardware. This is not a primary use-case for NVDA, which is why this issue has not been prioritized. As far as we are aware, the majority of users are served well by the current options.

There may be other use-cases for wanting more options for the NVDA key, but they aren't described well here. To be properly prioritized, they should be raised as new issues and described in detail. The title of this issue is too general, doesn't match the description well, I'll update it.

**Comment by CyrilleB79** [comment]
> There may be other use-cases for wanting more options for the NVDA key, but they aren't described well here. To be properly prioritized, they should be raised as new issues and described in detail. The title of this issue is too general, doesn't match the description well, I'll update it.

@feerrenrut, since you have updated the title and narrowed the scope of this ticket, I think that re-opening #7995 should be considered.
@mertdeg2 if you're still impacted by #7995, please comment there and re-open if needed.

I mention also that I have seen other requests in the past to change NVDA key, due to missing or unusual position of Insert key on innovative laptop keyboards. I won't open a new issue for this now since I do not have the precise example anymore; but will do if this request comes again.

**Comment by rdebeasi** [comment]
The "How to map your Mac's CapsLock key" article link above is broken. Here's an updated, working link: [How to map your Mac's CapsLock key to a NVDA or JAWS key in a Windows virtual machine](https://www.marcozehe.com/how-to-map-your-macs-capslock-key-to-a-nvda-or-jaws-key-in-a-windows-virtual-machine/)

**Comment by Adriani90** [comment]
Closing in favor of #16607 which contains a very valid use case and also the one related to the mac keyboard.
